packages/@sanity/form-builder/src/components/Details.tsx

```diff
@@ -1,50 +1,71 @@
-import {ChevronDownIcon} from '@sanity/icons'
-import React from 'react'
+import {ToggleArrowRightIcon} from '@sanity/icons'
+import {Box, Flex, Text} from '@sanity/ui'
+import React, {useCallback, useState} from 'react'
+import styled from 'styled-components'
 
-import styles from './Details.css'
-
-type DetailsProps = {
-  isOpen?: boolean
+interface DetailsProps {
+  children?: React.ReactNode
+  open?: boolean
   title?: React.ReactNode
 }
 
-type DetailsState = {
-  isOpen: any
-}
+const HeaderButton = styled.button`
+  -webkit-font-smoothing: inherit;
+  appearance: none;
+  font: inherit;
+  background: none;
+  width: 100%;
+  text-align: left;
+  border: 0;
+  margin: 0;
+  padding: 0;
+  outline: none;
+`
 
-export class Details extends React.Component<DetailsProps, DetailsState> {
-  static defaultProps = {
-    title: 'Details',
-    isOpen: false,
-  }
+const Header = styled(Flex)`
+  cursor: default;
+  user-select: none;
+`
 
-  constructor(props) {
-    super(props)
-    this.state = {
-      isOpen: props.isOpen,
-    }
+const IconBox = styled(Box)`
+  & > div > svg {
+    transform: rotate(0);
+    transition: transform 100ms;
   }
 
-  handleToggle = () => {
-    this.setState((prevState) => ({isOpen: !prevState.isOpen}))
+  &[data-open] > div > svg {
+    transform: rotate(90deg);
   }
+`
 
-  render() {
-    const {title, children} = this.props
-    const {isOpen} = this.state
-
-    return (
-      <div className={styles.root} data-open={isOpen}>
-        <button type="button" className={styles.headerButton} onClick={this.handleToggle}>
-          <div className={styles.header}>
-            <span className={styles.iconContainer}>
-              <ChevronDownIcon />
-            </span>
-            <span className={styles.summary}>{title}</span>
-          </div>
-        </button>
-        <div className={styles.content}>{children}</div>
-      </div>
-    )
-  }
+export function Details(props: DetailsProps) {
+  const {children, open: openProp, title = 'Details'} = props
+
+  const [open, setOpen] = useState(openProp || false)
```

Comment:
Can we assume it's controlled by the consumer If we receive a non-undefined `open` prop and not maintain internal state here? As it is now a consumer can't fully control its open/closed state.

Reply:
Fixed